designer/server/src/main/scala/pl/touk/nussknacker/ui/process/ProcessStateDefinitionService.scala

```diff
@@ -37,37 +34,40 @@ class ProcessStateDefinitionService(typeToConfig: ProcessingTypeDataProvider[Pro
           categoriesWhereStateAppears
         )
       }
-      .collect { case d if d.categories.nonEmpty => d }
+      .filter(_.categories.nonEmpty)
       .toList
   }
 }
 
-// TODO: Move validation and stateDefinitions aggregation to ProcessingTypeDataReader (where processing types are reloaded).
-// Now we have only late validation when UI asks for state definitions, and state aggregation is performed here
-// every time validation is executed, but should happen only once per configuration reload.
-// That requires ProcessingTypeDataProvider.all to be an entity that combines all processing types AND
-// has all-processing-type aggregates, such as aggregated stateDefinitions.
 object ProcessStateDefinitionService {
 
+  type StatusNameToStateDefinitionsMapping = Map[StatusName, (StateDefinitionDetails, List[ProcessingType])]
+
   /**
     * Each processing type define its own state definitions. Technically it is possible that two processing types provide
     * states with the same StatusName and different UI configurations (displayable name and icon). Here is an assertion
     * that this does not happen and each state has the same definition across all processingTypes.
     */
-  def checkUnsafe(processingTypes: Map[ProcessingType, ProcessingTypeData]): Unit = {
-    val countDefinitionsForName = processingTypeStateDefinitions(processingTypes)
+  def createDefinitionsMappingUnsafe(processingTypes: Map[ProcessingType, ProcessingTypeData]): StatusNameToStateDefinitionsMapping = {
+    import cats.instances.list._
+    import cats.syntax.alternative._
+
+    val validatedProcessingTypeStateDefinitions = processingTypeStateDefinitions(processingTypes)
       .groupBy { case (_, statusName, _) => statusName }
       .map { case (statusName, stateDefinitionsForOneName) =>
         val uniqueDefinitionsForName = stateDefinitionsForOneName
+          // TODO: Ask whether we should really aggregate by both displayable name and icon. Shouldn't it be only displayable name?
```

Comment:
@gskrobisz please have a look at it :slightly_smiling_face:

Reply:
Well, that's somewhat confusing and that's my fault ;)
We should avoid situation when more than one processing type specifies a status, e.g. statusName = RUNNING, with different displayableName (e.g. `Running` and `Running, but different`) or different icons. Otherwise it makes a problem with a display of status filter options. Hence the expecteation of full coherence of status name, displayable name and icon among all processing types.

Reply:
Maybe we should have just status, without state status.. ?